client/components/illustration/index.jsx

```diff
@@ -0,0 +1,21 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+
+export const Illustration = ( { width = 200, height, path, alt, className, ...otherProps } ) => {
+	const src = require( 'assets/images/illustrations/' + path );
```

Comment:
This is gonna fail hard if the path doesn't exist. But the same failure occurs if we use `import img from 'assets/images/illustration/this-image-does-not-exist.svg';` so maybe it's OK?

Reply:
Given we have more illustrations in other paths like `assets/customer-home` why don't we force to pass the full path so it works with any illustration regardless where it is located?

```
<Illustration path="assets/custom-folder/my-illustration.svg" width="400" alt="My illustration" />
```

Or are we planning to move all illustrations to the same folder? If we end up having them in multiple locations I wonder if we actually need a component for this, since it's only passing down the props to a `img` element.

Reply:
> Given we have more illustrations in other paths like assets/customer-home why don't we force to pass the full path so it works with any illustration regardless where it is located?

That's a possibility! I'd hoped to migrate all of them to one folder for easier reference -- so designers don't have to go hunting to find illustrations in the code, and we don't upload duplicates. We could have sub-folders within `/illustrations` (like `/illustrations/customer-home`) if we wanted to sort them further, but having everything in one folders means we can reuse them in different places without duplication.

TBH I started working on the `/devdocs` illustration page and realized it would be easier to have a component so I didn't have to do as much work to list out all the IMGs, widths, etc. :D

> Or are we planning to move all illustrations to the same folder? If we end up having them in multiple locations I wonder if we actually need a component for this, since it's only passing down the props to a img element.

My thought is this also ensures the image is automatically using webpack without having to do the imports separately.

Reply:
I'm torn on this. On one hand I totally get having them all in one place. But on the other hand my own mental model of assets is grouped by section (e.g. `assets/customer-home`).

Would it make sense to do a check on the path passed in the prop? Something like checking the path for `assets/` and if it doesn't exist then default to `assets/images/illustrations`?

Reply:
> I'm torn on this. On one hand I totally get having them all in one place. But on the other hand my own mental model of assets is grouped by section (e.g. assets/customer-home).

I hear that. My thought was to make the illustrations location agnostic so we can reuse them across the product. I wonder if grouping them by section is compensating for the lack of documentation/a single source of truth, which this PR aims to resolve?

> Would it make sense to do a check on the path passed in the prop? Something like checking the path for assets/ and if it doesn't exist then default to assets/images/illustrations? 

I haven't figured out how to check if a file exists yet, but I'm pretty sure it's possible. Another option is to look at `assets/images` and pass the rest of the path from there. Then we don't have to move files around but we're still restricted to the images directory, which is where (presumably?) all image assets would live.

Reply:
> My thought is this also ensures the image is automatically using webpack without having to do the imports separately.

Oh right, that's true 👍 

> I haven't figured out how to check if a file exists yet, but I'm pretty sure it's possible

`require` is a synchronous operation so you can just wrap it in a try/catch:

```js
try {
  const src = require( 'assets/images/illustrations/' + path );
  // Render image
} catch (ex) {
  // Handle error
}
```

> Another option is to look at `assets/images` and pass the rest of the path from there

Alternatively we can add support for an additional prop that indicates the category or section of the image.

Examples:

```
// Will retrieve the illustration from the default folder assets/images/illustrations
<Illustration path="my-illustration.svg" width="400" alt="My illustration" />
```

```
// Will retrieve the illustration from a grouping folder folder assets/images/customer-home
<Illustration path="my-illustration.svg" section="customer-home" width="400" alt="My illustration" />
```

Reply:
I'd +1 just passing through the full path for now. We can revisit in the future. 

It helps with code searches and I'm not sure we're consistent enough in Calypso to be worth adding this convention.

Reply:
I also haven't played around with this yet, but I think we may be able to use`await import` for dynamic imports.

```jsx
export default async function () {

if ( fileExists ) { //probably better as an assertion vs runtime check
   const asset = await import( `assets/images/illustrations/${ path }` );
   //... do other stuff.
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

Reply:
Hmm, this is an Interesting approach, and I'm not sure what the best way of going about this is.

So, I should point out that this isn't an actual import that's happening here. The file isn't loaded by that `require` statement; instead, `webpack` has some magic to convert that string into a different string, which includes the cache-busting hash for the asset. So I'm not sure what happens if we provide a filename that's missing and whether a `try`/`catch` would be the right way of handling that.

In fact, I'm not sure how well this will work at all, regardless of whether we use a `require` or a dynamic `import` as @gwwar suggested. We may have to include some webpack magic comments to get this to play nicely without, say, appending a large dictionary of all illustration URLs to this module.

Any idea what the right approach is here, @scinos, @griffbrad?

Reply:
I can confirm from my own testing that trying to require a file that's missing does *not* fail gracefully. :D But trying to import a file that does not exist explicitly (via `import img from 'path/to/non-existent-img.svg';`) fails the same way. 🤷‍♀️

Reply:
@jsnajdr: I know you've handled these sorts of imports and associated `webpack` magic in the past. Any idea what the right approach is here?